src/instrumentsscene/view/staffsettingsmodel.cpp

```diff
@@ -53,27 +60,44 @@ void StaffSettingsModel::load(const QString& staffId)
     emit voicesChanged();
     emit cutawayEnabledChanged();
     emit isSmallStaffChanged();
+    emit allStaffTypesChanged();
+    emit staffTypeChanged();
 }
 
 QVariantList StaffSettingsModel::allStaffTypes() const
 {
     QVariantList result;
+    auto staff = notationParts()->staff(m_staffId);
+    if (staff == nullptr) {
+        return result;
+    }
+    int maxLines = 0;
+    bool isPercussion = false;
+    if (staff->part() && staff->part()->instrument()) {
+        if (staff->part()->instrument()->stringData()) {
+            maxLines = staff->part()->instrument()->stringData()->frettedStrings();
+        }
+        isPercussion = staff->part()->instrument()->useDrumset();
+    }
 
-    for (notation::StaffType type: notation::allStaffTypes()) {
-        QVariantMap obj;
+    for (const Ms::StaffType& type : Ms::StaffType::presets()) {
+        if (isPercussion ? type.group() == Ms::StaffGroup::PERCUSSION
+            : type.group() == Ms::StaffGroup::STANDARD || type.group() == Ms::StaffGroup::TAB && type.lines() <= maxLines) {
```

Comment:
This condition looks too complicated to read/debug. You should at least move it to temporary variables